### ISSUE TITLE
[Feature] Add page.waitForPageEvent() function

### ISSUE THREAD

**Issue by rightaway** [microsoft/playwright#15440]
page.waitForEvent() confusingly doesn't wait for page events https://github.com/microsoft/playwright/issues/6220.

The workaround is to use page.evaluate() https://github.com/microsoft/playwright/issues/6220#issuecomment-822669293 but there's no way to set a timeout for this so if the event never happens it blocks forever. And the syntax to do that is not simple and understandable.

It would be better to add a new function that waits for a page event such as `page.waitForPageEvent("my-event")` or add a new parameter to the existing waitForEvent function such as `page.waitForEvent("my-event", { page: true })`.

**Comment by rightaway** [comment]
If there's some app logic that isn't straightforward that you need playwright to wait for, sometimes the only way to do it is to dispatch your own custom event and wait for it. It's a useful pattern like `renderAfterDocumentEvent` in https://github.com/chrisvfritz/prerender-spa-plugin and other frameworks.

It would be nice to have a simple `page.waitForPageEvent("my-event")` or `page.waitForEvent("my-event", { page: true })` since the workarounds in the link or your suggestion are not expressive for what it's trying to do which is just wait for a custom event. Such function could be better than other wait functions in certain cases.

**Comment by aslushnikov** [comment]
I see. Yes, this pattern does exist, but we don't see it too much out in the wild. Those few that do use it rely on very different signaling methods:
* some use [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent)
* some use `console.debug` messages
* some use `console.timestamp` messages
* some use framework events

As a result, everybody ends up with their own utility method to await such a signal.

**Comment by rightaway** [comment]
They're using all these methods because there's no function in playwright to wait on a standard dom event (https://developer.mozilla.org/en-US/docs/Web/API/Event). Waiting on console.debug and console.timestamp messages is a workaround.

If someone looked in playwright documentation and saw `waitForDomEvent` function they wouldn't need to use those methods and write their own Promise.race which is not at all clear like `waitForDomEvent("my-event")`.

A function that lets you wait on a dom event would give developers a standard and simple api to use to wait for something custom. Many frameworks have this feature because it makes development easy and developers choose it over the other options that try to duplicate this.

**Comment by MarioGruda** [comment]
Some frameworks that support hydration emit an event as soon as it is complete.

For example, SvelteKit: https://kit.svelte.dev/docs/events

I ended up with the following to wait for hydration and it works, but it would be nice to have a documented built in function in playwright

```js
await page.evaluate(eventName => new Promise(callback => window.addEventListener(eventName, callback, { once: true })), 'sveltekit:start');
```

**Comment by alexbjorlig** [comment]
> Some frameworks that support hydration emit an event as soon as it is complete.
> 
> For example, SvelteKit: https://kit.svelte.dev/docs/events
> 
> I ended up with the following to wait for hydration and it works, but it would be nice to have a documented built in function in playwright
> 
> ```js
> await page.evaluate(eventName => new Promise(callback => window.addEventListener(eventName, callback, { once: true })), 'sveltekit:start');
> ```

I have been using this method, but I sometimes don't get the event if the `page.goto` is loading a very small page that is pre-rendered.

The `sveltekit:start` event is there, but I think the problem is that it's emitted before the eventListener is added.

**Comment by JonMarbach** [comment]
+1 I am new to Playwright and I also assumed waitForEvent would wait for a DOM event. The suggestions here for how to achieve it don't feel very "natural" or maybe a better way to put it is that they don't feel native to Playwright (e.g., they don't interact with Playwright's built-in timeout values.)

**Comment by iansan5653** [comment]
I also just ran in to this. I want to wait for something that doesn't change the DOM, and it's not a network request. So I figured the cleanest way to avoid a timeout was to call `dispatchEvent(new CustomEvent('MyEventName'))` and then I expected to be able to call `page.waitForEvent('MyEventName')`.

An alternative to a new method might be a new overload on `waitForEvent`, like `page.waitForEvent('domevent', {name: 'MyEventName'})` (or the predicate form `page.waitForEvent('domevent', e => e.name === 'MyEventName')`).

**Comment by ITenthusiasm** [comment]
I would argue that being able to listen for DOM events is a basic and necessary feature for tools testing web applications. I've been trying out Playwright while experimenting with Web Components. And overall, I really like it. (I especially like using `await` over using Cypress closures, though that comes with its own challenges.) But I do think this is an important feature that's missing, and I'd love to see it. Justification below if more feedback is needed:

## Justification for the OP

If you're using [Web Components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) to create an enhanced [select](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select) element that [participates in forms](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/attachInternals), there are a few things you'll want to test.

- You'll want to verify that the [`input`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) and [`change`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event) events get dispatched when the component's value changes. (In addition to mimicking the `select` element's behavior, this enables devs to take advantage of event delegation for forms, which is more performant and more maintainable than attaching listeners to every single form control.)
- You'll want to verify that -- just as with the `select` element -- pressing the `Enter` key will [`submit`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event) the form when the combobox is closed. You'll also want to verify that event was dispatched [with the correct default `submitter`](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#implicit-submission). (This provides a clear and familiar experience for both devs and end users.)

These are just 2 simple use cases for a component that's required in _many_ web applications. (People may not be using Web Components for their "custom select elements" yet, but they will still have similar requirements whether they're using Pure JS or a JS Framework.) Several other common-use-case components require tests that can listen for DOM events. So this is a critical feature for a library that tests web applications.

## Is This Use Case Common ?

> Yes, this pattern does exist, but we don't see it too much out in the wild.

I could be wrong... But I would wager that the [rarity of this pattern's existence](https://github.com/microsoft/playwright/issues/15440#issuecomment-1179017307) might be for 2 primary reasons:

1) **The Distribution of Tests**. Although testing for dispatched events is a rudimentary requirement for web applications, it's _less_ common than testing _everything else_ that pertains to web applications. Here's an example: I'm working on a custom Select Component that adheres to the [WAI ARIA guidelines](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/). Thus, _the large majority_ of my [tests](https://github.com/ITenthusiasm/custom-element-discoveries/blob/main/src/Combobox/__tests__/Combobox.test.ts) don't require listening for events. Instead I'm checking, "Did the correct `aria-*` attributes get updated when a user did such and such?" And that is indeed how most tests should be written: Test User Experience, not Implementation Details. However, components _do_ still have an API aspect to them that needs to be tested. If I'm going to have something that replaces the `select` element, then -- as I mentioned earlier -- it _needs_ to dispatch the `input` and `change` events when its value changes, and it _needs_ to be able to submit forms. However, the tests for these use cases are going to be _much smaller_ in size and in quantity than the other tests. That may explain the scarcity, but it doesn't change the fact that those tests are _incredibly_ important for proving that the component behaves as needed.
2) **Playwright's Age**. Playwright is still pretty new, and it's competing with Cypress + Testing Library and Jest/Vitest + Testing Library. It's quite possible that this use case isn't seen as frequently because enough people haven't moved to Playwright yet. It's also possible that some people _aren't_ committing to Playwright because this basic need isn't satisfied yet.

I know it's not much, but in 5 years of software development with various projects, I have not yet come across an application/project that _did not_ need to test the dispatching of specific DOM events in some way. I assume similar things could be said for frontend developers who've worked longer in the field -- particularly those who've built components. Again, this is a very common use case for components.

## Playwright vs. Cypress and Jest/Vitest

In Jest/Vitest, verifying that an event was dispatched is very straightforward:
```ts
const handler = jest.fn(); // Or vi.fn();
document.querySelector("form").addEventListener("submit", handler);

// After performing some user interactions ...
expect(handler).toHaveBeenCalledWith(expect.objectContaining({ /* ... Important properties, such as `submitter` ... */ });
```

[Cypress](https://glebbahmutov.com/blog/listen-to-events/) also enables you to listen for events from the DOM. This is very important if you're using Web Components, because [JSDOM](https://github.com/jsdom/jsdom) does not support important features like `elementInternals` yet.

But as far as I can tell, Playwright has no idiomatic way to handle these requirements. What's worse, Playwright has no _easy_, _reliable_ way to handle these requirements. Currently, [to test that a `<form>` was submitted](https://github.com/ITenthusiasm/custom-element-discoveries/blob/next/src/Combobox/__tests__/Combobox.test.ts#L748-L760), I have to register an event handler with the form that updates a `data-*` attribute. This, of course, is brittle and prone to error. So I end up having to lay down specific rules: "_This constant_ is the attribute that you should be testing for", "If you're only testing form submissions, only use `defaultSubmissionHandler`", etc. The more I try to make things reusable and clear, the more complex abstractions there are; and all of these abstractions are still less safe because what I'm testing (an element has a specific attribute) is farther away from what I'm _really_ trying to test (an event listener was properly called).

I've never gotten caught up in such complexity with Cypress or Jest/Vitest.

## Final Comments

Usually, when a developer has to bend over backwards to accomplish something basic (as in the scenario I just described), it's an indicator that the testing tool is lacking something. Another Example: When writing [Jest](https://jestjs.io/) tests for a library that had non-trivial asynchronous logic, I didn't have a way to see whether a `Promise` returned from a mock function had settled or not. So I ended up having to [extend](https://stackoverflow.com/questions/30564053/how-can-i-synchronously-determine-a-javascript-promises-state/76838887#76838887) and replace the native `window.Promise` to test what I needed. Because [Vitest](https://vitest.dev/api/mock.html#mock-results) satisfied this need out of the box with their mock functions, it was convenient (and safer) to switch to their tool rather than stick to my own hacky way of solving the problem in Jest.

**Comment by NiklasBeierl** [comment]
I second @ITenthusiasm's take and want to share another workaround.

I wanted to emit a `CustomEvent` via an init_script and then catch and further process the event in my playwright code.
 
However `page.on("my-custom-event", callback)` never triggered. As a workaround, I can offer hooking up your event to handler that makes an http request and then intercepting that request via playwrights `page.route`  or `page.waitForRequest`. The js handler converting the event to an http call (serializing the event, etc) can be injected as an init_script and probably be made generic enough to re-use it for any kind of event.

**Comment by pavelfeldman** [comment]
Closing to be transparent in that we are not going to implement it. We understand the general interest in the issue, but there are different interpretations of it and overall lack of interest (<20 upvotes over 2.5 years).

**Comment by ITenthusiasm** [comment]
I'm working on a `combobox` Web Component that's ergonomic and Framework Agnostic.

After running into a situation where I needed to verify that various events were being fired off with the correct data, I created a helper function to make tracking DOM Events in Playwright less tedious.

- [Helper Function](https://github.com/ITenthusiasm/custom-element-discoveries/blob/next/src/Combobox/__tests__/Combobox.test.ts#L6115-L6152) (Search for `trackEvents` if the link is incorrect.)
- [Example Usage](https://github.com/ITenthusiasm/custom-element-discoveries/blob/next/src/Combobox/__tests__/Combobox.test.ts#L6154-L6183) (Search for "event when the user selects a new" if the link is incorrect.)

Hope this helps some of you.
